File: source/17.电话号码的字母组合.cpp
/*
 * @lc app=leetcode.cn id=17 lang=cpp
 *
 * [17] 电话号码的字母组合
 */
#include "cpp_includes.h"
// @lc code=start
class Solution {
public:
    string keyboard[10] = {"", "", "abc", "def", "ghi", "jkl", "mno", "pqrs", "tuv", "wxyz"};
    vector<string> letterCombinations(string digits) {
        vector<string> res;
        string str;

        if (digits.empty()) return res;

        letterCombinations(res, str, digits, 0);
        return res;
    }

    void letterCombinations(vector<string>&res, string str, string digits, int index) {
        if (str.length() == digits.length()) {
            res.push_back(str);
            return;
        }

        string keys = keyboard[digits[index] - '0'];
        for (char c : keys) {
            letterCombinations(res, str+c, digits, index+1);
        }
    }
};
// @lc code=end

File: source/17.电话号码的字母组合.cs
/*
 * @lc app=leetcode.cn id=17 lang=csharp
 *
 * [17] 电话号码的字母组合
 */
using csharpnaming;
// @lc code=start
public class Solution
{
  private List<string> keyboard = new List<string>() {
      "", "", "abc", "def", "ghi", "jkl",
      "mno", "pqrs", "tuv", "wxyz" };
  public IList<string> LetterCombinations(string digits)
  {
      if (digits=="") return new List<string>();
      var res = new List<string>();
      string str = "";

      LetterCombinations(res, str, digits, 0);
      return res;
  }

  private void LetterCombinations(List<string> res, string str, string digits, int index) {
      if (index == digits.Length) {
          res.Add(str);
          return;
      }

      string chars = keyboard[digits[index] - '0'];
      foreach(var c in chars) {
          LetterCombinations(res, str+c, digits, index+1);
      }
  }
}
// @lc code=end

